packages/transaction-controller/src/utils/simulation.ts

```diff
@@ -477,24 +477,47 @@ function getEventTokenIds(event: ParsedEvent): (Hex | undefined)[] {
 }
 
 /**
- * Extract the value from a balance transaction response.
+ * Extract the value from a balance transaction response using the correct ABI.
  * @param from - The address to check the balance of.
  * @param token - The token to check the balance of.
  * @param response - The balance transaction response.
- * @returns The value of the balance transaction.
+ * @returns The value of the balance transaction as Hex.
  */
-function getValueFromBalanceTransaction(
+export function getValueFromBalanceTransaction(
   from: Hex,
   token: SimulationToken,
   response: SimulationResponseTransaction,
 ): Hex {
-  const normalizedReturn = normalizeReturnValue(response.return);
+  const interfaces = getContractInterfaces();
+  let decoded;
 
-  if (token.standard === SimulationTokenStandard.erc721) {
-    return normalizedReturn === from ? '0x1' : '0x0';
+  try {
+    if (token.standard === SimulationTokenStandard.erc721) {
+      // ERC721: Check ownership and return 1 if the owner is the user, 0 otherwise.
+      decoded = (
+        interfaces.get(SupportedToken.ERC721) as Interface
+      ).decodeFunctionResult('ownerOf', response.return);
+      const owner = decoded[0].toLowerCase();
+      return owner === from.toLowerCase() ? '0x1' : '0x0';
+    }
+    // Non-ERC721: ERC20 and ERC1155 use balanceOf
+    decoded = (
+      interfaces.get(
+        token.standard === SimulationTokenStandard.erc20
+          ? SupportedToken.ERC20
+          : SupportedToken.ERC1155,
+      ) as Interface
+    ).decodeFunctionResult('balanceOf', response.return);
+
+    return toHex(decoded[0] as Hex);
```

Comment:
Why do we need `toHex` if it's already `Hex`?

Reply:
Done.